### PR TITLE
Make spawn egg colours visible

### PR DESCRIPTION
**Pull request by luisc99**
Adds an unsafe method for getting the underlying colour definitions for spawn eggs, based on the entity type. I recognise this is probably a fairly niche thing, but it can be useful to have a consistent way of getting some colours for entities when doing certain visual effects.

I've added this to UnsafeValues as it's accessing data directly from the SpawnEggItem class rather than reading data from an individual ItemStack's meta, although I'm happy for it to be moved somewhere more appropriate if desired.

The code is based on ItemFactory#getSpawnEgg, [as found in this patch](https://github.com/PaperMC/Paper/blob/a80d31b1bd84a56f969634c39d78dfa1c437aa9f/patches/server/0655-Add-ItemFactory-getSpawnEgg-API.patch).

### CODE REVIEW

File: patches/api/0458-Make-spawn-egg-color-visible.patch [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
++    /**
++     * Obtains the underlying color informating for a spawn egg of a given
++     * entity type, or -1 if the entity passed does not have a spawn egg.
++     * Spawn eggs have two colors - the background layer (0), and the
++     * foreground layer (1)
++     * @param entityType The entity type to get the color for
++     * @param layer The texture layer to get a color for
++     * @return The color of the layer for the entity's spawn egg
++     */
++    int getSpawnEggLayerColor(org.bukkit.entity.EntityType entityType, int layer);
```
Comment (yannicklamprecht) [inline patches/api/0458-Make-spawn-egg-color-visible.patch]: Does the int represent a RGB or RGBA color value or is it just a index for an enum?
Please either clarify in JavaDoc or return a `@Nullable Color` instead.

**Review by yannicklamprecht — CHANGES_REQUESTED** [review]

File: patches/api/0458-Make-spawn-egg-color-visible.patch (hunk 1 above)
Comment (kennytv) [reply patches/api/0458-Make-spawn-egg-color-visible.patch]: Yeah it should be a Color object, not the raw int

**Comment by yannicklamprecht** [comment]
Please squash commits in each submodule and rebuildPatches.

**Comment by luisc99** [comment]
Whoops, didn't notice it made a new patch, fixed

**Review by yannicklamprecht — APPROVED** [review]

**Review by Warriorrrr — APPROVED** [review]
I think it might be nice to add getters for the background & highlight color to spawn egg meta, but currently also lgtm

File: patches/server/1060-Make-spawn-egg-color-visible.patch [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
++    // Paper start - spawn egg color visibility
++    @Override
++    public org.bukkit.Color getSpawnEggLayerColor(org.bukkit.entity.EntityType entityType, int layer) {
++        String typeId = entityType.getKey().toString();
++        net.minecraft.resources.ResourceLocation typeKey = new net.minecraft.resources.ResourceLocation(typeId);
++        net.minecraft.world.entity.EntityType<?> nmsType = net.minecraft.core.registries.BuiltInRegistries.ENTITY_TYPE.get(typeKey);
++        net.minecraft.world.item.SpawnEggItem eggItem = net.minecraft.world.item.SpawnEggItem.byId(nmsType);
++        return eggItem == null ? null : org.bukkit.Color.fromRGB(eggItem.getColor(layer));
++    }
++    // Paper end
```
Comment (yannicklamprecht) [inline patches/server/1060-Make-spawn-egg-color-visible.patch]:
```suggestion
+    // Paper end - spawn egg color visibility
```

**Review by yannicklamprecht — CHANGES_REQUESTED** [review]

File: patches/api/0458-Make-spawn-egg-color-visible.patch [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
++     * Obtains the underlying color informating for a spawn egg of a given
++     * entity type, or null if the entity passed does not have a spawn egg.
++     * Spawn eggs have two colors - the background layer (0), and the
++     * foreground layer (1)
++     * @param entityType The entity type to get the color for
++     * @param layer The texture layer to get a color for
++     * @return The color of the layer for the entity's spawn egg
++     */
++    @Nullable org.bukkit.Color getSpawnEggLayerColor(org.bukkit.entity.EntityType entityType, int layer);
++    // Paper end
```
Comment (yannicklamprecht) [inline patches/api/0458-Make-spawn-egg-color-visible.patch]:
```suggestion
+    // Paper end - spawn egg color visibility
```
Same here.

**Review by yannicklamprecht — CHANGES_REQUESTED** [review]

**Review by Machine-Maker — APPROVED** [review]
If this wasn't in Unsafe, I would have some comments, but I think since it is, and therefore API compat isn't guaranteed, I think its fine. 

Perhaps some more stable API can be added once we have subtypes of ItemType and BlockType in the API.

**Pull request merged**